core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -152,32 +150,31 @@ protected boolean resolveIndex() {
         IndexRequest indexRequest = (IndexRequest) item.request();
         preVersions[requestIndex] = indexRequest.version();
         preVersionTypes[requestIndex] = indexRequest.versionType();
-        try {
-            WriteResult<IndexResponse> result = shardIndexOperation(request, indexRequest, metaData, indexShard, true);
-            location = locationToSync(location, result.getLocation());
-            // add the response
-            IndexResponse indexResponse = result.getResponse();
-            setResponse(item, new BulkItemResponse(item.id(), indexRequest.opType().lowercase(), indexResponse));
-        } catch (Exception e) {
-            // rethrow the failure if we are going to retry on primary and let parent failure to handle it
-            if (retryPrimaryException(e)) {
+        WriteResult<IndexResponse> result = shardIndexOperation(request, indexRequest, metaData, indexShard, true);
+        if (result.operationFailed()) {
+            Exception failure = result.getFailure();
+            if (retryPrimaryException(failure)) {
                 // restore updated versions...
                 for (int j = 0; j < requestIndex; j++) {
                     applyVersion(request.items()[j], preVersions[j], preVersionTypes[j]);
                 }
-                throw (ElasticsearchException) e;
             }
-            logFailure(e, "index", request.shardId(), indexRequest);
+            logFailure(failure, "index", request.shardId(), indexRequest);
             // if its a conflict failure, and we already executed the request on a primary (and we execute it
             // again, due to primary relocation and only processing up to N bulk items when the shard gets closed)
             // then just use the response we got from the successful execution
-            if (item.getPrimaryResponse() != null && isConflictException(e)) {
+            if (item.getPrimaryResponse() != null && isConflictException(failure)) {
                 setResponse(item, item.getPrimaryResponse());
             } else {
                 setResponse(item, new BulkItemResponse(item.id(), indexRequest.opType().lowercase(),
-                        new BulkItemResponse.Failure(request.index(), indexRequest.type(), indexRequest.id(), e)));
+                    new BulkItemResponse.Failure(request.index(), indexRequest.type(), indexRequest.id(), failure)));
             }
+            return null;
         }
+        location = locationToSync(location, result.getLocation());
```

Comment:
can we put this first under result.operationFailed() == false? I think it will be easier to read for most people (they want to see the "normal" flow and it's shorter? als we don't run the risk of forgetting a return statement because the compiler will tell us